app/services/charts/colors_array.rb

```diff
@@ -0,0 +1,58 @@
+class Charts::ColorsArray < Struct.new(:count)
+  def colors
+    collors_array
+  end
+
+  private
+
+  def collors_array
+    golden_ratio_conjugate = 0.618033988749895
+    h = rand
+
+    gen_html do
+      h += golden_ratio_conjugate
+      h %= 1
+      hsv_to_rgb(h, 0.6, 0.85)
+    end
+  end
+
+  def gen_html
+    result = []
+    count.times do
+      r, g, b = yield
+      result << "rgb(#{r}, #{g}, #{b})"
+    end
+    result
+  end
+
+  def hsv_to_rgb(h, s, v)
+    h_i = (h * 6).to_i
+    f = h * 6 - h_i
+    p = v * (1 - s)
+    q = v * (1 - f * s)
+    t = v * (1 - (1 - f) * s)
+
+    hsv_to_rgb_values_converter(h_i, p, q, t, v)
+  end
+
+  def hsv_to_rgb_values_converter(h_i, p, q, t, v)
+    case h_i
+    when 0
+      set_rgb(v, t, p)
+    when 1
+      set_rgb(q, v, p)
+    when 2
+      set_rgb(p, v, t)
+    when 3
+      set_rgb(p, q, v)
+    when 4
+      set_rgb(t, p, v)
+    when 5
+      set_rgb(v, p, q)
+    end
+  end
+
+  def set_rgb(r, g, b)
+    [(r * 256).to_i, (g * 256).to_i, (b * 256).to_i]
+  end
+end
```

Comment:
I wanted to improve this service but I don't know how I can do it

Reply:
@marderer as a fast refactor I can see:
```ruby
class Charts::ColorsArray < Struct.new(:count)
  def collors_array
    h = rand

    gen_html do
      h += golden_ratio_conjugate
      h %= 1
      hsv_to_rgb(h, 0.6, 0.85)
    end
  end

  private

  def gen_html
    count.times.inject do |result|
      r, g, b = yield
      result << "rgb(#{r}, #{g}, #{b})"
    end
  end

  def hsv_to_rgb(h, s, v)
    Charts::HSVtoRGB.new(h, s, v).convert
  end


  def golden_ratio_conjugate
    0.618033988749895
  end
end

class Charts::HSVtoRGB < Struct.new(h, s, v)
  def convert
    h_i = (h * 6).to_i
    f = h * 6 - h_i
    p = v * (1 - s)
    q = v * (1 - f * s)
    t = v * (1 - (1 - f) * s)

    hsv_to_rgb_values_converter(h_i, p, q, t, v)
  end

  private

  def hsv_to_rgb_values_converter(h_i, p, q, t, v)
    case h_i
    when 0
      set_rgb(v, t, p)
    when 1
      set_rgb(q, v, p)
    when 2
      set_rgb(p, v, t)
    when 3
      set_rgb(p, q, v)
    when 4
      set_rgb(t, p, v)
    when 5
      set_rgb(v, p, q)
    end
  end

  def set_rgb(r, g, b)
    [(r * 256).to_i, (g * 256).to_i, (b * 256).to_i]
  end
end
```

Reply:
@marderer I can't understand what is `h` and `f`, `p`, `q`, `t`. Give them normal names.

Reply:
@AleksSenkou I wrote about variables in [slack](https://datarockets-school.slack.com/archives/C40MCAB0V/p1522931458000401?thread_ts=1522930971.000048&cid=C40MCAB0V)

Reply:
@AleksSenkou it's just the variables that I use for storing calculations, if I change the names sense wont change.